### PR TITLE
Gradio-lite (Gradio Wasm)

### PR DESCRIPTION
**Pull request by whitphx**
# Description

A Wasm-port of Gradio using Pyodide.

* `/wasm` directory contains the Wasm-related code. The architecture is like Shinylive or stlite where Pyodide runs on a WebWorker (`src/webworker/index.ts`) and the frontend app communicates with the server running on the worker through a "worker proxy" object (`src/worker-proxy.ts`).
* The frontend app (`@gradio/app`) in the `js/app/` directory uses the worker proxy.
  The frontend app is using a client library (`@gradio/client` in the `client/js` directory), so I added `overridden_fetch` option to the client library to inject the `fetch()` and used it to make the client object to dispatch requests to the worker proxy.
    * At this point, I choose this design because it minimizes the changes to the existing code of the frontend app and the client, while I don't think it's the best in some aspects, e.g. we should define a plugable interface to inject the HTTP communication layer instead of directly inject the `fetch()` function. So let us discuss about how we should modify the design around the client to switch between the normal HTTP comms (`fetch`) and the emulated HTTP for Wasm.
* As written above, currently, the Wasm-version is directly implemented modifying the code in `@gradio/app` and it will be compiled into the `/lite` directory with `pnpm build:lite` command.
  
TODO:
* [x] Fix the build not to affect the normal build of `@gradio/app` (making dead code elimination to work on `webworker.js`).
  * Done with b379f09. All imports of `@gradio/wasm` have been removed from `Index.svelte` and lifted up to `lite/index.ts`, which is an entry point for the Wasm ver but not used for the normal build.
  * It looks like Vite's dead code elimination doesn't work for some dynamic imports including our worker initializations (`new Worker(new URL("./webworker.js", import.meta.url));`) even if this code is not actually used in the final build. So moving all related imports to the parent file that is built only for the Wasm ver. was a solution.
    * Ref: https://github.com/vitejs/vite/issues/5676

While there are still many things to improve, let me suggest this PR at this point as a minimum working implementation.

The current build is deployed at https://huggingface.co/spaces/whitphx/gradio-static-test so that you can try it.
The `dist` directory in this repo is a copy of the build which you can get by running `pnpm build:lite` in the `/js/app` directory whose output would be `/lite/dist`.

### Code details:
Mainly I added/edited the following packages:
* `/wasm`: As written above, Pyodide runs on a WebWorker. This package contains the code for the WebWorker (`/wasm/src/webworker/*`) and a "WorkerProxy" class (`/wasm/src/worker-proxy`), which is used to communicate with the WebWorker from the main thread.
  All WebWorker-related implementation is encapsulated in this package, and I believe this package doesn't have any side effects on main part of the Gradio frontend app.
* `/gradio`: I would like the reviewers to check the changes in this package as they affects the existing Gradio Python package.
    * I added `wasm_utils.py`
        * to check if the Python code is running in a Wasm env (`is_wasm` attribute),
        * and to expose the Gradio's ASGI app object (the `app` object of the `FastAPI` instance) with `register_app` and `get_registered_app`.
    * I edited `blocks.py` to initialize the `app` object **without the ASGI server (uvicorn)**.
    * Other changes are mainly to stop importing or using modules which are not available in Pyodide.
* `/js/app`
    * I reused `Index.svelte` and most parts of the build configs (`vite.config.js`) because it's already well-developed and I wanted to keep the existing code unchanged as much as possible.
    * I also added another entry point `lite.html` and `lite/index.ts` for the Wasm version. There are used instead of `index.html` when we run `pnpm build:lite` or `pnpm dev:lite`. I also added some build configs for them.
    * Some needed to intercept some external resource accesses via HTTP such as `fetch()` or `mount_css()` to make them access the resources served from the WebWorker. So I made `Index.svelte` export `mount_css` and `overridden_fetch` to inject these functions from `lite/index.ts` for Wasm.
* `/client/js`: This is also what I would like the reviewers to check especially because, as written above, I updated the client module to override the `fetch()` function to access the resources served from the WebWorker.

Future Works
* ~~Make `File` and `UploadButton` work on Wasm, which uses `upload_files`.~~ Done 9284911
* Pre-build the CSS to apply the styles to the loading screen before the server initialization is done.
* Reload the app without re-initialization of the worker.
* Distribute the library via CDN.
* Support FFMpeg.
* 
Closes: # (issue)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added a short summary of my change to the CHANGELOG.md
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

**Comment by abidlabs** [comment]
Taking a look at the Python changes now. Btw I removed the changelog entry since I believe this is not meant to be a user-facing change right now

File: gradio/wasm_utils.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import sys
+
+# See https://pyodide.org/en/stable/usage/faq.html#how-to-detect-that-code-is-run-with-pyodide
+is_wasm = sys.platform == "emscripten"
+is_pyodide = "pyodide" in sys.modules
```
Comment (abidlabs) [inline gradio/wasm_utils.py:5]: Suggest naming these as constants
```suggestion
IS_WASM = sys.platform == "emscripten"
IS_PYIODIDE = "pyodide" in sys.modules
```

Comment (abidlabs) [reply gradio/wasm_utils.py:5]: Also, is `is_pyiodide` used at all?

File: gradio/components.py [hunk 2]
```diff
@@ -74,6 +73,10 @@
 from gradio.interpretation import NeighborInterpretable, TokenInterpretable
 from gradio.layouts import Column, Form, Row
 
+if not wasm_utils.is_wasm:
+    # TODO: Support ffmpeg on Wasm
+    from ffmpy import FFmpeg
```
Comment (abidlabs) [inline gradio/components.py:78]: Probably would be good to raise an error in `gr.Video()` where `FFmpeg` is used  explaining that certain features are not available in wasm.

**Comment by abidlabs** [comment]
Python side looks good! Just left a few nits, see above

File: gradio/components.py (hunk 2 above)
Comment (whitphx) [reply gradio/components.py]: I added `wasm_utils.WasmUnsupportedError` and made `gr.Video` to raise it when ffmpeg is to be used 👍

File: gradio/wasm_utils.py (hunk 1 above)
Comment (whitphx) [reply gradio/wasm_utils.py]: 👍

**Comment by whitphx** [comment]
@abidlabs Thanks, I first modified the Python parts.

**Comment by whitphx** [comment]
@pngwn Thanks, I updated as follows:
* Create `api_factory()` to create the functions that use `fetch()` including not only `client()` but also `post_data()` and `upload_files`.
* Move `/wasm` to `/js/wasm`.
* Move `/lite` to `/js/lite`.
    * I think this package will be published as `gradio/lite` after adding some necessary files e.g. `package.json`, so left its name as is.

**Comment by whitphx** [comment]
During the change on `client`, I also made `upload_files` Wasm-compatible and updated the components using it. 6a0288f

**Comment by whitphx** [comment]
@abidlabs @pngwn 
(This is not a main topic though, just a feedback)

> Btw I removed the changelog entry when merging the latest changes from main since I believe this is not meant to be a user-facing change right now

I agreed it, but also got confused feeling the policy looked inconsistent, for example I was asked to write the changelog for internal updates such as https://github.com/gradio-app/gradio/pull/4222.

**Comment by abidlabs** [comment]
Thanks @whitphx for making the changes! Regarding the last point, actually my bad -- Thinking about it some more, it's worth adding refactoring changes to the changelog because developers of the codebase may want to know in which version of Gradio certain changes were introduced, particularly for a refactor as large as this one. However, I wouldn't put it in the "new features" section of the changelog, since that's intended mostly for users of the library. I would put it in the "other changes" section for now.

**Comment by whitphx** [comment]
@abidlabs All makes sense, and it's clear! Thanks :)

**Comment by whitphx** [comment]
Note: the conflict on `gradio/components.py` is because it has been split into each file per component and deleted in #4487.
So it seems to be ok if we just apply the changes in this PR to the new file, `video.py`.

**Review by pngwn — APPROVED** [review]
I'm happy with this. There are some things it would be good to change but those things depend on some pending refactors, so we can take care of them later.

When a few more people have tested this doesn't break anything then we can get this in.

Thanks @whitphx! Awesome work!

**Review by abidlabs — APPROVED** [review]
LGTM! All the demos look good to me + re-reviewed the Python changes.

**Comment by dawoodkhan82** [comment]
Tested some of the demos, lgtm.

**Review by dawoodkhan82 — APPROVED** [review]

**Comment by julien-c** [comment]
yay! exciting stuff

**Pull request merged**